【功能模块开发】任务调度/评测机池

fitenne opened the pull request:
Fixes #10 

评测机池
  添加 login、logout、getToken

评测队列
  修改 push 参数为 JudgeRequest

添加调度、评测队列文档

添加调试用接口 login、logout

Commit:
评测机池 基本代码

评测队列 基本代码

Commit:
评测机池
  添加 login、logout、getToken

评测队列
  修改 push 参数为 JudgeRequest

添加调度、评测队列文档

添加调试用接口 login、logout

删除关于缓存的代码

Commit:
评测机池
  添加 login、logout、getToken

评测队列
  修改 push 参数为 JudgeRequest

添加调度、评测队列文档

添加调试用接口 login、logout

删除关于缓存的代码

A0nameless0man commented on src/scheduler/judge-queue-service/judge-queue-service.service.ts:
```diff
@@ -0,0 +1,32 @@
+     */
+    async push(judgeRequest: JudgeRequest): Promise<string> {
+        await this.redisService.withClient(async client => {
+            return client.rpush(
```
这个return有意义吗……

fitenne replied:
已删除。

Commit:
为 judger-pool-service 中 getTokenReleaser 的返回值添加类型声明。

删除了 judge-queue-service 中无意义的 return。

合并 SchedulerService 中连续的 await。

Handgrip commented on src/scheduler/judge-queue-service/judge-queue-service.service.ts:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * push a JudgeRequest to queue, return it's taskid
+     */
+    async push(judgeRequest: JudgeRequest): Promise<string> {
```
push进来的应该是taskid，而没有详细内容

Handgrip commented on src/scheduler/judge-queue-service/judge-queue-service.service.ts:
```diff
@@ -0,0 +1,32 @@
+        return judgeRequest.taskId;
+    }
+
+    async pop(): Promise<JudgeRequest> {
```
这里应该也是taskid

Handgrip commented on src/scheduler/judge-queue-service/judge-queue-service.service.ts:
```diff
@@ -0,0 +1,32 @@
+     * push a JudgeRequest to queue, return it's taskid
+     */
+    async push(judgeRequest: JudgeRequest): Promise<string> {
+        await this.redisService.withClient(async client => {
```
```ts
        await this.redisService.withClient(async client => 
            await client.lpush(
                JudgeQueueService.pendingQueue,
                JSON.stringify(judgeRequest)
            );
        );
// 或者
        await this.redisService.client.lpush(...)
```

Handgrip commented on src/scheduler/judge-queue-service/judge-queue-service.service.ts:
```diff
@@ -0,0 +1,32 @@
+     */
+    async push(judgeRequest: JudgeRequest): Promise<string> {
+        await this.redisService.withClient(async client => {
+            client.rpush(
```
左push，右pop

Handgrip commented on src/scheduler/judge-queue-service/judge-queue-service.service.ts:
```diff
@@ -0,0 +1,32 @@
+    async pop(): Promise<JudgeRequest> {
+        return await this.redisService.withClient(async client => {
+            const ret = (
+                await client.blpop(JudgeQueueService.pendingQueue, 0)
```
左push，右pop

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,72 @@
+        this.logger.log(
+            `tring to login: ${judgerId} with capacity of ${capacity}`
+        );
+        for (let i = 0; i < capacity; ++i) {
```
```ts
let mu = this.redisService.client.multi();
for(...;...;...){
    mu=mu.sadd(...).lpush(...);
}
await mu.exec();

fitenne replied:
已修改

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,72 @@
+                    token
+                )) == 1
+            ) {
+                this.redisService.client
```
await

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    async getToken(): Promise<[string, () => Promise<void>]> {
+        await this.redisService.withClient(async client => {
```
```ts
        await this.redisService.withClient(async client => {
            return await client.blpop(JudgerPoolService.tokenCount, "0");
        });
or 
        await this.redisService.withClient(async client => 
            await client.blpop(JudgerPoolService.tokenCount, "0");
        );
```
左push，右pop

blpop第二个参数好像是timeout吧

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,72 @@
+        });
+        const token = (await this.redisService.client.spop(
+            JudgerPoolService.tokenBucket
+        )) as string;
```
判 undefined

fitenne replied:
```ts
        await this.redisService.withClient(async client => {
            return client.brpop(JudgerPoolService.tokenCount, 0);
        });
        const token = (await this.redisService.client.spop(
            JudgerPoolService.tokenBucket
        )) as string;
```
每个 token 对应 tokenCount 中一个元素。逻辑上此处不会得到 undefined。
得到 undefiend 要怎么处理？

Handgrip replied:
【如果】碰巧哪里出了bug，能拦住。
throw new Error("xxx");

Handgrip replied:
其实应该是判null
```ts
if(!token) throw new Error("xxx");
```

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,72 @@
+
+    async logout(judgerId: string): Promise<number> {
+        this.logger.log(`loging out: ${judgerId}`);
+        for (let i = 0; ; ++i) {
```
若1235号令牌均已归还，4号未归还，则5号令牌不会被清理。

A0nameless0man replied:
这里redis是否提供了通配符之类的东西？或者考虑上lua脚本？

Nugine replied:
不主动清理令牌，当获取令牌时丢弃无效的，这样如何？

A0nameless0man replied:
也是一个方案

fitenne replied:
我喜欢 Nugine 的方法。
关于 “获取令牌时丢弃无效的”， @Flaryer 评测机状态在 Redis JudgerReport 中吗，我可以直接查询一下状态？
这样对评测机池来说，评测机注销就简化成了 “评测机交互” 部分直接改变一下评测机状态。

Handgrip replied:
同一评测机所有令牌的值一样可以吗（令牌放在队列里，可重复，失去了随机性），比如就是评测机id的值。另外令牌放队列里可以少维护一个tokenCount。这样release的时候就不用携带token或者调用回调函数了（懒） @Nugine @A0nameless0man @fitenne

Nugine replied:
> 同一评测机所有令牌的值一样可以吗（令牌放在队列里，可重复，失去了随机性），比如就是评测机id的值。另外令牌放队列里可以少维护一个tokenCount。这样release的时候就不用携带token或者调用回调函数了（懒） @Nugine @A0nameless0man @fitenne

我觉得没什么影响，可以这样简化。

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,72 @@
+        token: string
+    ): Promise<() => Promise<void>> {
+        return async (): Promise<void> => {
+            this.redisService.client
```
期望上要检测评测机是否还在评测机池内，否则会归还失效令牌。

fitenne replied:
[重定向](https://github.com/ThinkSpiritLab/heng-controller/pull/14#discussion_r567596065)

Handgrip commented on src/scheduler/scheduler.debug.controller.ts:
```diff
@@ -0,0 +1,32 @@
+    async createJudge(
+        @Body() createJudgeRequest: CreateJudgeRequest
+    ): Promise<string> {
+        return this.judgeQueue.push(createJudgeRequest.body.mainJudge);
```
await

Handgrip commented on src/scheduler/scheduler.debug.controller.ts:
```diff
@@ -0,0 +1,32 @@
+
+    @Post("/test/JudgerPool/logout")
+    async logout(@Body() info: any): Promise<string> {
+        this.JudgerPool.logout(info["name"]);
```
await

Handgrip commented on src/scheduler/scheduler.debug.controller.ts:
```diff
@@ -0,0 +1,32 @@
+
+    @Post("/test/JudgerPool/login")
+    async login(@Body() info: any): Promise<string> {
+        this.JudgerPool.login(info["name"], info["maxTaskCount"]);
```
await

A0nameless0man reviewed (changes requested):
请将 @Flaryer 指出的部分修正，同时请记得修改文档。

Nugine commented on src/scheduler/scheduler.debug.controller.ts:
```diff
@@ -0,0 +1,32 @@
+
+    @Post("/test/judgeQueue/push")
+    async createJudge(
+        @Body() createJudgeRequest: CreateJudgeRequest
```
应当使用 dto 验证 post 参数。参考 [类验证器](https://docs.nestjs.cn/7/pipes?id=%e7%b1%bb%e9%aa%8c%e8%af%81%e5%99%a8)。

Nugine commented on src/scheduler/scheduler.debug.controller.ts:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Post("/test/JudgerPool/login")
+    async login(@Body() info: any): Promise<string> {
```
最好避免写出 any，这里应当使用 dto，如果仅用来测试，可以原地定义一下参数类型。例如 
```ts
info: { name: string, maxTaskCount: number }
```

fitenne replied:
仅用来测试，使用了后者。

Nugine commented:
主分支已更新，基于 master 分支 rebase 一下。

A0nameless0man commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,72 @@
+        };
+    }
+
+    async login(judgerId: string, capacity: number): Promise<number> {
```
这里要不要同时传入Software信息？以后可能能修改算法支持特定语言评测机？

fitenne replied:
这个要怎么样？

Nugine replied:
@A0nameless0man 以后要改的时候再想办法吧

Handgrip commented on src/scheduler/scheduler.service.ts:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    async run(): Promise<void> {
+        while (true) {
```
循环内部需要加try catch，保证循环不会中断

fitenne replied:
Re-request review

fitenne committed:
统一`评测队列`用法为左 push 右 pop。
修改`评测队列`只保存 taskid。

获取 token 时确认获取到了 token。

修改了 debug 使用的 url，现与 Redis 调试使用的 url 格式保持一致。
scheduler.debug.controller.ts 若干 bug 修复。

scheduler.service 中的循环使用 try catch。

文档更新

fitenne committed:
统一`评测队列`用法为左 push 右 pop。
修改`评测队列`只保存 taskid。

获取 token 时确认获取到了 token。

修改了 debug 使用的 url，现与 Redis 调试使用的 url 格式保持一致。
scheduler.debug.controller.ts 若干 bug 修复。

scheduler.service 中的循环使用 try catch。

fitenne committed:
Merge remote-tracking branch 'scheduler/master'

fitenne committed:
修改了调度算法

修改 JudgeQueueService JudgerPoolService 若干返回值为 Primose<void>

修改了文档

A0nameless0man commented on src/scheduler/scheduler.service.ts:
```diff
@@ -0,0 +1,45 @@
+    async run(): Promise<void> {
+        while (true) {
+            try {
+                const [taskId, token] = await Promise.all([
```
解构赋值可以嵌套

```js
const [taskId, [token,releaser]] = await Promise.all...
```

A0nameless0man replied:
感觉这样写也不太美观……
@Nugine 大佬怎么看？

fitenne replied:
修改了调度算法，现在不需要 `releaser` 了

A0nameless0man commented on src/scheduler/scheduler.service.ts:
```diff
@@ -0,0 +1,45 @@
+                    ", send to",
+                    judger,
+                    "wait for 5 sec..."
+                );
```
可以使用反引号拼接字符串

fitenne replied:
这个是未写完的，不打算放到最后的代码里啊。

A0nameless0man replied:
了解了

A0nameless0man reviewed (changes requested):
请将 `console.log` 替换为 `Logger`

fitenne committed:
添加之前忘记添加的 exports 字段。

Handgrip commented on src/scheduler/judge-queue-service/judge-queue-service.service.ts:
```diff
@@ -0,0 +1,25 @@
+
+    async pop(): Promise<string> {
+        return await this.redisService.withClient(async client => {
+            return (await client.brpop(JudgeQueueService.pendingQueue, 0))[1];
```
阻塞监听的timeout不应该是0，这里可能要设置为10s，30s之类的，外面套一个while，判断是否成功取出了任务，没取出就重新监听。

A0nameless0man replied:
我测试了一下，删除一个队列不能使其上的阻塞监听返回，甚至包括将其删除后将键值设为其它类型也不行。

A0nameless0man replied:
不过只要这个队列没有惨遭删除的可能性，应该问题不大

Handgrip replied:
😥好像一样的，目的都是一直监听，好像不改也行）

fitenne replied:
[blpop#blocking-behavior](https://redis.io/commands/blpop#blocking-behavior)

```
If none of the specified keys exist, BLPOP blocks the connection until another client performs an LPUSH or RPUSH operation against one of the keys.

A timeout of zero can be used to block indefinitely.😄️
```

Nugine replied:
我觉得这里改不改都行

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,82 @@
+        let ret: [string, string] | undefined;
+        while (true) {
+            ret = await this.redisService.withClient(async client => {
+                return client.brpop(JudgerPoolService.tokenBucket, 0);
```
同上

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,82 @@
+            ret = await this.redisService.withClient(async client => {
+                return client.brpop(JudgerPoolService.tokenBucket, 0);
+            });
+            if (ret === undefined) {
```
if(!ret)

A0nameless0man replied:
我其实更倾向于这种语义比较清晰的写法…………
@Nugine 大佬怎么看？

Handgrip replied:
好像不是undefined，而是null

Flaryer replied:
搞错了）确实是undefined😓

Handgrip replied:
null

fitenne replied:
fixed

fitenne committed:
修正了 getToken 中 token 的类型错误。

Nugine commented on src/scheduler/scheduler.debug.controller.ts:
```diff
@@ -0,0 +1,37 @@
+import { JudgeQueueService } from "./judge-queue-service/judge-queue-service.service";
+import { ExternalProtocol } from "heng-protocol";
+import { JudgerPoolService } from "./judger-pool/judger-pool.service";
+import CreateJudgeRequest = ExternalProtocol.Post.CreateJudgeRequest;
```
协议仓库已更新，现在可以用最新的定义了。

<https://github.com/ThinkSpiritLab/Heng-Protocol/pull/6>

Nugine commented:
各位检查一下还有没有要改的。

PR 作者改完后及时 Resolve conversation，reviewer 如果要求进一步改动，可以 Unsolve conversation。如果讨论之后一致同意不需要改，由 reviewer 及时 Resolve conversation。

明天我将合并此 PR，请 @Flaryer  #16  做好对接准备。

可以提前把分支拉下来测试有没有合并冲突

```bash
git checkout -b fitenne-master master
git pull https://github.com/fitenne/heng-controller.git master
```

fitenne committed:
更新了 Heng-Protocol 和以及相关使用

Nugine commented:
PR 作者改完后在对应的讨论里面回复一下 fixed by 1ce96b9 之类的，然后再 resolve，这样看邮件找上下文会方便一些。

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,82 @@
+
+    async getToken(): Promise<string> {
+        let ret: [string, string] | null;
+        while (true) {
```
有个问题，执行redis命令时是有可能抛出错误的（如键的类型和命令接受的类型不同、redis断联），那么这个错误会被抛到getToken外面去（因为这里没有拦截）。
这里调用的时候外面（run函数）加了异常拦截，但是以后调用在其他地方会不会出问题。
所以我感觉有while的地方都要加异常拦截👀

Handgrip replied:
再想想问题好像也不是非常大，因为之前加异常拦截是因为while不能断，这里没这个需求🤐

Handgrip commented on src/scheduler/judger-pool/judger-pool.service.ts:
```diff
@@ -0,0 +1,82 @@
+                throw new Error("[judger pool]获取 token 失败");
+            }
+            if (
+                await this.redisService.client.sismember(
```
但是假如这里异常了，这个令牌就丢失了
